In [1]:
import pandas as pd

In [2]:
energy = pd.read_csv('energydata_complete.csv')
energy.head(2)

date  Appliances  lights     T1       RH_1    T2     RH_2  \
0  2016-01-11 17:00:00          60      30  19.89  47.596667  19.2  44.7900   
1  2016-01-11 17:10:00          60      30  19.89  46.693333  19.2  44.7225   

      T3   RH_3    T4  ...         T9   RH_9     T_out  Press_mm_hg  RH_out  \
0  19.79  44.73  19.0  ...  17.033333  45.53  6.600000        733.5    92.0   
1  19.79  44.79  19.0  ...  17.066667  45.56  6.483333        733.6    92.0   

   Windspeed  Visibility  Tdewpoint        rv1        rv2  
0   7.000000   63.000000        5.3  13.275433  13.275433  
1   6.666667   59.166667        5.2  18.606195  18.606195  

[2 rows x 29 columns]

In [3]:
energy.shape

(19735, 29)

In [4]:
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import Ridge
from sklearn.linear_model import Lasso
from sklearn.metrics import mean_squared_error, mean_absolute_error

In [5]:
X = energy['T2'].values.reshape(-1, 1) #Input Variaable
y = energy['T6'] #Target variable i.e. the label

In [6]:
#Training Linear Regression Model
linear_reg = LinearRegression()
linear_reg.fit(X,y)

LinearRegression()

## Q17. What is the Root Mean Squared error in three D.P?

In [7]:
import numpy as np
pred = linear_reg.predict(X)
mse = mean_squared_error(pred, y)
rmse = np.sqrt(mse)
rmse

3.6444644873484573

**Answer: 3.644**

In [8]:
X = energy.drop(['date', 'lights', 'Appliances'], axis = 1) #creating a new input with some columns dropped
y = energy['Appliances'] #Target

In [9]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
X_scaled = scaler.fit_transform(X)

In [10]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y , test_size= 0.3, random_state= 42)

In [11]:
# Training Linear Regression Model
linear_reg1 = LinearRegression()
linear_reg1.fit(X_train, y_train)

LinearRegression()

## Q18. What is the Mean Absolute Error (in three decimal places) for the  training set?

In [12]:
# Model Evaluation on the train set
y_train_preds = linear_reg1.predict(X_train)
mae = mean_absolute_error(y_train, y_train_preds)
mae

53.74623381332399

**Answer: 53.746**

## Q19. What is the Root Mean Squared Error (in three decimal places) for the training set?

In [13]:
y_preds = linear_reg1.predict(X_train)
mse = mean_squared_error(y_train, y_preds)
rmse = np.sqrt(mse)
rmse

95.21557103245476

**Answer: 95.216**

## Q20. What is the Mean Absolute Error (in three decimal places) for test set?

In [14]:
y_preds = linear_reg1.predict(X_test)
mae = mean_absolute_error(y_test, y_preds)
mae

53.645886006282495

**Answer: 53.646**

## Q21. What is the Root Mean Squared Error (in three decimal places) for test set?

In [15]:
y_preds = linear_reg1.predict(X_test)
mse = mean_squared_error(y_test, y_preds)
rmse = np.sqrt(mse)
rmse

93.63775000857623

**Answer: 93.638**

## Q22. Did the Model above overfit to the training set

**Answer: No, because the RMSE value on the test set is smaller compare to the train set.**

## Q23. Train a ridge regression model with default parameters. Is there any change to the root mean squared error (RMSE) when evaluated on the test set?

In [16]:
# Training Ridge Regression Model
ridge = Ridge()
ridge.fit(X_train, y_train)

Ridge()

In [17]:
# Ridge Model Evaluation on the test set Using MSE
y_preds = ridge.predict(X_test)
ridge_mse = mean_squared_error(y_test, y_preds)
ridge_rmse = np.sqrt(ridge_mse)
ridge_rmse

93.71565336757126

**Answer:Yes**

## Q24. Train a lasso regression model with default value and obtain the new feature weights with it. How many of the features have non-zero feature weights?

In [18]:
lasso = Lasso()
lasso.fit(X_train, y_train)

Lasso()

In [19]:
lasso.coef_

array([  0.        ,  37.49194311,   0.        ,  -0.        ,
         0.        ,   0.        ,  -0.        ,   0.        ,
        -0.        ,   0.        ,   0.        ,  -0.        ,
        -0.        ,  -0.        ,   0.        , -11.4994144 ,
        -0.        ,  -0.        ,   0.        ,  -0.        ,
       -52.13992414,   5.62458594,   0.        ,   0.        ,
        -0.        ,  -0.        ])

**Answer: 4**

## Q25. What is the new RMSE with the Lasso Regression on the test set?

In [20]:
y_preds = lasso.predict(X_test)
lasso_mse = mean_squared_error(y_test, y_preds)
lasso_rmse = np.sqrt(lasso_mse)
lasso_rmse

99.81294734273483

**Answer: 99.813**